### ISSUE TITLE
GPU Hessian for Quantile regression

### ISSUE THREAD

**Issue by david-waterworth** [catboost/catboost#732]
I'm closing in on getting Quantile regression working very well on catboost...

I've found one last issue. The Quantile loss function has no 2nd derivative. Catboost returns 1.0 as the sample 2nd derivative. This means newtons method f'(x) / f''(x) collapses to gradient descent f'(x) **except** we're summing the samples in each leaf, so the result for Quantile regression is we're maximising the -ve loss function, but the gradient is averaged in each leaf and convergence is extremely slow (N times slower than it should be where N is the average samples per leaf).

I verified this by disabling the check for `--leaf-estimation-iterations` for Quantile loss function. In the test example below it takes around 100 leaf iterations to converge. I verified that for 20 samples, 1 split, 10 samples per leaf the update was 1/10th of what it should be for gradient descent. 

Then I hacked the code to force f''(x) = 1 for the leaf hessian's (I actually divided the leaf hessain by the number of leafs per tree as that was simpler than figuring out how to do it properly). 

I then got excellent convergence (with a small amount of L2 normalisation) with a small number of steps - in general with the tweaks I've made the results appear to be similar to lightgbm

I've also found starting from the mean helps

I'll try and create a WIP pull request to try and get this integrated.

```
import numpy as np
from catboost import CatBoostRegressor
from catboost import Pool

X = np.array([1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2], dtype=np.float32)
y = np.array([1,2,3,4,5,6,7,8,9,10,21,22,23,24,25,26,27,28,29,30], dtype=np.float32)
alpha=0.75

train_data = Pool(X, y)
a0 = np.mean(y)
train_data.set_baseline(np.repeat(a0,len(y)))

model = CatBoostRegressor(task_type='GPU',l2_leaf_reg=0.5,leaf_estimation_backtracking='Armijo',leaf_estimation_iterations=5,learning_rate=1,iterations=2,loss_function=f'Quantile:alpha=0.75',boosting_type='Plain',logging_level='Verbose')
model.fit(train_data);

model.predict(X)+a0
```

**Comment by olveirap** [comment]
Did you try starting from the median? It should be a more natural starting point.

**Comment by david-waterworth** [comment]
@olveirap I tried a range of start points. I found it didn't really matter as long as you start from somewhere in the middle rather than the edge of the distribution, and it's less effective than I hoped. The ``best`` point is probably the mean of the solution over the leaves which is a little bit above both the mean and median (and not known in advance). But the main issue seems to be the way the 2nd derivative is applied when loss == quantile 

i.e. with y0=0, the sample derivatives f'(x)=0.75 and the split is at 1.5 so sum f'(x) for both leaves is 7.5

The problem is catboost seems to set f''(x)=1 for every leaf, so sum f''(x) = 10

and the newton update is f'(x) / f''(x)=7.5/10   [I'm ignoring l2 normalisation which reduces f''(x)]

because f''(x) isn't the true derivative, it's only being set to 1 since the quantile loss function has no 2nd derivative.

I believe catboost is trying to fall back to gradient descent in this case - but by summing rather than averaging the derivatives and the loss function this doesn't actually work in practise?

Then on top of that, performing more than 1 leaf iteration is only supported for categorical not regression loss functions.

I'm hoping someone from the catboost team can confirm / deny this? I'd like to continue to use ``catboost`` as although I find for my regression model it's a lot slower than ``lightgbm``, it also finds a much better solution. But I need to estimate the mean and quantiles and I don't want to use two libraries.

**Comment by annaveronika** [comment]
We currently support Gradient method and Exact method for Quantile, Exact is set by default. You can use more than one step for Gradient method. And we've implemented starting from best constant, to use it please set boost_from_average=True (we use this name to be consistent with other libs, and will add a synonym to that).
Thank you for your valuable input on Quantile regression!

**Comment by david-waterworth** [comment]
Thanks @annaveronika I can confirm this makes a huge difference to convergence performance!
